single-pool/js/packages/client/src/instructions.ts

```diff
@@ -0,0 +1,327 @@
+      ReadonlyAccount<typeof SYSVAR_STAKE_HISTORY_ID>,
+      ReadonlyAccount<typeof STAKE_CONFIG_ID>,
+      ReadonlyAccount<typeof SYSTEM_PROGRAM_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type DepositStakeInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      WritableAccount<PoolStakeAddress>,
+      WritableAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolStakeAuthorityAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      WritableAccount<Base58EncodedAddress>, // user stake
+      WritableAccount<Base58EncodedAddress>, // user token
+      WritableAccount<Base58EncodedAddress>, // user lamport
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof SYSVAR_STAKE_HISTORY_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type WithdrawStakeInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      WritableAccount<PoolStakeAddress>,
+      WritableAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolStakeAuthorityAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      WritableAccount<Base58EncodedAddress>, // user stake
+      WritableAccount<Base58EncodedAddress>, // user token
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type CreateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      WritableSignerAccount<Base58EncodedAddress>, // mpl payer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+      ReadonlyAccount<typeof SYSTEM_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type UpdateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<VoteAccountAddress>,
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      ReadonlySignerAccount<Base58EncodedAddress>, // authorized withdrawer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+enum SinglePoolInstructionType {
+  InitializePool,
+  DepositStake,
+  WithdrawStake,
+  CreateTokenMetadata,
+  UpdateTokenMetadata,
+}
+
+export class SinglePoolInstruction {
```

Comment:
If we make this a class, despite it being non-instantiatable and having no internal state, it means that if you want _one_ instruction among these you _have_ to download them all.

If you split this into individually exported functions (eg. `createInitializePoolInstruction()`) then a dev can import one at a time and have their build system dead code eliminate the rest.

Reply:
sad, i really like having namespaces...

Reply:
`createCreateTokenMetadataInstruction` `createUpdateTokenMetadataInstruction` :nauseated_face:

Reply:
oh im a genius, i can keep my classes and also provide the individual functions lol

Reply:
oh i guess you cant define them like that but this is all syntax sugar anyway so i can just provide an object lol